lib/mpl_toolkits/tests/test_mplot3d.py

```diff
@@ -64,6 +65,29 @@ def test_bar3d_notshaded():
     fig.canvas.draw()
 
 
+def test_bar3d_lightsource():
+    fig = plt.figure()
+    ax = fig.add_subplot(1, 1, 1, projection="3d")
+
+    ls = mcolors.LightSource(azdeg=0, altdeg=90)
+
+    length, width = 3, 4
+    area = length * width
+
+    x, y = np.meshgrid(np.arange(length), np.arange(width))
+    x = x.ravel()
+    y = y.ravel()
+    dz = x + y
+
+    color = [cm.coolwarm(i/area) for i in range(area)]
+
+    collection = ax.bar3d(x=x, y=y, z=0,
+                          dx=1, dy=1, dz=dz,
+                          color=color, shade=True, lightsource=ls)
+
+    np.testing.assert_array_equal(color, collection._facecolors3d[1::6])
```

Comment:
What I don't understand here is, if the facecolors are actually the same as the colors taken from the colormap, are we really testing that the lightsource works? That is, if there is a lightsource, shouldn't the colors be *different*  from the ones from the coolwarm colormap?

Reply:
`._facecolors3d[1::6]` should be taking the color of the top surfaces, which with a lightsource coming from the top (`altdeg=90`), should leave these unchanged (fully illuminated), as expected.

Changing the altdeg (e.g. default value of `19.4712`) or selecting different sides (e.g. `.facecolors3d[2::6]`) fails this test.

A test could be made which computes to expected colors given different angles, or compares to an image, but I thought this simple approach would suffice, as it ensures that the default lightsource is not used.

It's worth mentioning that the test would also pass if no shading were done, but `test_bar3d_shaded` should already test for this.

Reply:
I see. So imagine in the future someone unintentionally removes `lightsource` from the line `self._shade_colors(facecolors, normals, lightsource)`. The test would not detect this, right? But that is what we have tests for.

Reply:
The test would detect this.

Calling `self._shade_colors(facecolors, normals)` without a lightsource argument defaults to `LightSource(azdeg=225, altdeg=19.4712)`, which would fail the test.

Reply:
Oh, that is not very obvious. I'd suggest to add that as a comment in the test; else I fear in the future people might be as confused as I was by it.

Reply:
May I suggest the following as comment in the test:

```
# Testing that the custom 90° lightsource produces different shading than the default,
# and that those colors are precisely the colors from the colormap, due to the
# illumination parallel to the z-axis.
```

Reply:
Thanks, that sounds much better. I also added that it compares the top facecolors.
```
#(...) different shading on the top facecolors compared to the default (...)
```